### ISSUE TITLE
decode: 2 low-severity audit findings (wave-1 batch)

### ISSUE THREAD

**Issue by forkwright** [forkwright/logismos#51]
Migrated from `forkwright/logismos-archive#72`, because the archived tracker is not watched. Both findings were re-verified against live `main`; `crates/decode/src/processors.rs` is byte-identical between the archived tree and live `main`, so the cited lines are unchanged and both findings are still present.

Kept as one batch issue rather than two micro-issues, per the batch's original framing.

## Finding

Two low-severity findings in the `decode` crate's `TopK` logits processor: a NaN logit bypasses the filter entirely, and the processor allocates a vocabulary-sized scratch buffer on every decode step.

## Evidence

**correctness**

- `crates/decode/src/processors.rs:41` — the NaN-safe sort treats `NaN` as `Equal`, which avoids a panic but does not filter it. A `NaN` logit always fails the `*l < threshold` comparison, so it is never masked regardless of where the sort places it:

  ```rust
  sorted.sort_by(|a, b| b.partial_cmp(a).unwrap_or(core::cmp::Ordering::Equal));
  ```

**resilience**

- `crates/decode/src/processors.rs:40` — a full vocab-sized `Vec` is allocated on every `process()` call, i.e. every decode step:

  ```rust
  let mut sorted: Vec<f32> = logits.to_vec();
  ```

## Why this matters

A `NaN` that survives top-k filtering propagates into the sampling step, where it can select an arbitrary token or poison the distribution. `NaN` logits are reachable from upstream defects already tracked in this repo (the all-masked-row softmax path), so this is the second line of defence failing rather than a purely theoretical concern.

The per-step allocation is pure overhead on the decode hot loop — a vocabulary-sized copy per token, at 32k–150k `f32` for typical vocabularies.

## Desired correction

- Handle `NaN` explicitly in `TopK`: either mask `NaN` logits to `NEG_INFINITY` before sorting, or reject them with an error rather than letting them pass the threshold comparison.
- Hoist the scratch buffer into the processor so it is allocated once and reused across decode steps.

Done when: a `NaN` logit is either masked or surfaced as an error rather than bypassing the top-k filter, and `TopK::process` does not allocate a vocab-sized buffer per call.
